server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorFactory.java

```diff
@@ -52,9 +52,18 @@ public Aggregator createInternal(SearchContext searchContext, Aggregator parent,
      * Create the {@linkplain Aggregator} for a {@link ValuesSource} that
      * doesn't have values.
      */
-    protected abstract Aggregator createUnmapped(SearchContext searchContext,
+    protected Aggregator createUnmapped(SearchContext searchContext,
                                                  Aggregator parent,
-                                                 Map<String, Object> metadata) throws IOException;
+                                                 Map<String, Object> metadata) throws IOException
+    {
+        return new NonCollectingAggregator(name, searchContext, parent, factories, metadata) {
+
+            @Override
+            public InternalAggregation buildEmptyAggregation() {
+                return new InternalNonMapped(name, metadata);
```

Comment:
I suspect this is the source of most of your test failures.  We still need to return the `InternalAggregation` type appropriate to that aggregation class.  I hadn't really thought about that when writing this ticket, and looking at it now I'm not sure we can implement it correctly here on `ValuesSourceAggregationFactory`.  In the spirit of incremental improvement, I suggest we leave `createUnmapped` abstract in `ValuesSourceAggregationFactory` for now, and in each of the aggregations you are changing, we can create an instance of `NonCollectingAggregator` which returns the appropriate `InternalAggregation` type here.

That at least lets us get rid of the `null` handling in the aggregators, which was the main objective here.  Once we have that in place, we can see if a more natural refactoring suggests itself.  I need to give some thought to it.  Long term, it feels like we have too many layers here (Builder -> Factory -> Supplier), and maybe as we work towards cleaning that up, a more natural refactoring will suggest itself.